File: FTP_Client.py
import socket
import os
import sys

#Send a command to the server and wait for a response.
def send_command(control_socket, command):
    control_socket.sendall(command.encode())
    response = control_socket.recv(1024).decode()
    return response

#ls command.
def handle_ls(control_socket):
    response = send_command(control_socket, "ls")
    print("Server response:\n", response)

#get command to download a file.
def handle_get(control_socket, filename):
    response = send_command(control_socket, f"get {filename}")
    if response.startswith("FAIL"):
        print("Server error:", response)
        return

    #Establish a data connection
    data_port = int(response)
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as data_socket:
        data_socket.connect((server_name, data_port))
        with open(filename, "wb") as file:
            while True:
                data = data_socket.recv(1024)
                if not data:
                    break
                file.write(data)
    print(f"File '{filename}' downloaded successfully.")

#put command to upload a file.
def handle_put(control_socket, filename):
    if not os.path.exists(filename):
        print(f"File '{filename}' not found.")
        return

    response = send_command(control_socket, f"put {filename}")
    if response.startswith("FAIL"):
        print("Server error:", response)
        return

    #Create a data connection
    data_port = int(response)
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as data_socket:
        data_socket.connect((server_name, data_port))
        with open(filename, "rb") as file:
            while chunk := file.read(1024):
                data_socket.sendall(chunk)
    print(f"File '{filename}' uploaded successfully.")


#Main FTP client.
def ftp_client(server_name, server_port):
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as control_socket:
        control_socket.connect((server_name, server_port))
        print("Connected to FTP server.")
        while True:
            command = input("ftp> ").strip()
            if command == "quit":
                send_command(control_socket, "quit")
                print("Disconnected from the server.")
                break
            elif command == "ls":
                handle_ls(control_socket)
            elif command.startswith("get "):
                filename = command.split(" ", 1)[1]
                handle_get(control_socket, filename)
            elif command.startswith("put "):
                filename = command.split(" ", 1)[1]
                handle_put(control_socket, filename)
            else:
                print("Invalid command. Supported commands: ls, get <filename>, put <filename>, quit.")

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python FTP-Client.py <server_name> <server_port>")
        sys.exit(1)

    server_name = sys.argv[1]
    server_port = int(sys.argv[2])
    ftp_client(server_name, server_port)

File: ftp-client.mjs
// Prompt the user
import fs from "node:fs";
import { createConnection, createServer } from "node:net";
import { stdout } from "node:process";
const serverName = process.argv[2];
const port = process.argv[3];

if (serverName == null || port == null) {
  console.error("Error: Please use node <server_name> <port>");
  process.exit(1);
}

console.log({ serverName, port });

// create control socket and connect to ftp server
const client = createConnection(port, serverName, () => {
  console.log("initial control socket port: ", client.address().port);
  console.log(`connection sucessfully established to ${serverName}:${port}`);

  process.stdout.write("ftp>");

  //receive cli input
  process.stdin.on("data", (data) => {
    const input = data.toString().trim();
    const [cmd, ...args] = input.split(" ");
    if (cmd == "quit" || cmd == "q" || cmd == "exit") {
      console.warn("Exiting...");
      process.exit(0);
    }

    console.log({ cmd, args });
    if (cmd == "ls") {
      handleLs(cmd, args);
    } else if (cmd == "put") {
      handlePut(cmd, args);
    } else if (cmd == "get") {
      handleGet(cmd, args);
    } else {
      process.stdout.write("ftp>");
    }
  });
});

function handleLs(cmd, args) {
  const dataChannel = createServer((socket) => {
    console.log(`${cmd}: data channel created`);

    socket.on("data", (chunck) => {
      const fileList = chunck.toString().split(",");
      console.table(fileList);
    });

    socket.on("end", () => {
      console.log("ls data channel ended");
      dataChannel.close();
    });
    socket.on("error", (err) => {
      console.error("Error in ls data channel:", err.message);
    });
  });
  dataChannel.on("close", () => {
    console.log("ls data channel closed");
    process.stdout.write("ftp>");
  });

  // Listen to unassigned port and send info to the FTP server
  dataChannel.listen(0, () => {
    const addr = dataChannel.address();
    const port = addr.port;
    console.log(typeof port);

    client.write(`${cmd} ${port}`, "utf-8");
    console.log("Data channel listening on port ", port);
  });
}

function handlePut(cmd, args) {
  const fileName = args[args.length - 1];
  const filePath = `./client-files/${fileName}`;
  try {
    const fileBuffer = fs.readFileSync(filePath);
    const dataChannel = createServer((socket) => {
      console.log("PUT data channel created");

      socket.write(fileBuffer, (err) => {
        if (err != undefined) {
          console.error(`${cmd} command error`, err.message);
          socket.destroy();
          dataChannel.end();
          throw err;
        } else {
          console.log(`${cmd}: SUCCESS`);
          socket.end();
          dataChannel.close();
        }
      });

      socket.on("end", () => {
        console.log(`${cmd}: data channel ended`);
        dataChannel.close();
      });
      socket.on("error", (err) => {
        console.error(`Error ${cmd} ls data channel:`, err.message);
        throw err;
      });
    });
    dataChannel.on("close", () => {
      console.log(`${cmd}: data channel closed`);
      process.stdout.write("ftp>");
    });

    // Listen to unassigned port and send info to the FTP server
    dataChannel.listen(0, () => {
      const addr = dataChannel.address();
      const port = addr.port;

      client.write(`${cmd} ${fileName} ${port}`);
      console.log(`${cmd}: data channel listening on port `, port);
    });
  } catch (error) {
    console.error(`${cmd} Error: `, error.message);
    process.stdout.write("ftp>");
  }
}

function handleGet(cmd, args) {
  const fileName = args[args.length - 1];
  const filePath = `./client-files/${fileName}`;
  let writeStream;
  try {
    const dataChannel = createServer((socket) => {
      socket.on("data", (chunck) => {
        const header = chunck.slice(0, 7).toString(); // check if fail header was sent
        // check if server sent an err msg
        if (header.startsWith("@@FAIL:")) {
          console.error(`${cmd}: server error: ${chunck}`);
          dataChannel.close();
          return;
        }
        console.log(chunck);

        // init writeStream once in case of multiple chuncks
        if (!writeStream) {
          writeStream = fs.createWriteStream(filePath);
        }

        writeStream.write(chunck, (err) => {
          if (err) {
            console.error(`${cmd}: data channel chunck error ${err.meessage}`);
          }
        });
      });

      socket.on("end", () => {
        console.log(`${cmd}: data channel ended`);
        if (writeStream) {
          writeStream.end();
        }
        dataChannel.close();
      });
      socket.on("close", (err) => {
        if (err) {
          console.error(`${cmd}: FAIL`);
        }
        if (writeStream) {
          writeStream.end();
        }
      });
      socket.on("error", (err) => {
        console.error(`Error ${cmd} get data channel:`, err.message);
      });
    });

    dataChannel.on("close", () => {
      console.log(`${cmd}: data channel closed`);
      process.stdout.write("ftp>");
    });

    // handle err if server DC etc
    dataChannel.on("error", (err) => {
      console.error(`${cmd}: FAIL: ${err.message}`);
    });

    // Listen to unassigned port and send info to the FTP server
    dataChannel.listen(0, () => {
      const addr = dataChannel.address();
      const port = addr.port;

      client.write(`${cmd} ${fileName} ${port}`);
      console.log(`${cmd}: data channel created on port `, port);
    });
  } catch (error) {
    console.error(`${cmd} Error: `, error.message);
    process.stdout.write("ftp>");
  }
}

client.on("error", (err) => {
  console.error(`Error: client socket error ${err.message}`);
});

client.on("close", () => {
  console.error(`Can't reach server`);
});

client.on("data", (data) => {});
